walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,106 @@
+   * @param request The meeting request that contains the requirements for potential meetings.
+   * @return A Collection of the feasible meeting {@code TimeRange}s.
+   */
+  private Collection<TimeRange> getMeettingTimes(List<Event> eventList, MeetingRequest request) {
+    eventList.add(
+        new Event(
+            "EOD", TimeRange.fromStartDuration(TimeRange.END_OF_DAY, 0), request.getAttendees()));
+    int endOfEarlierEvent = TimeRange.START_OF_DAY;
+
+    Collection<TimeRange> openMeetingTimes = new ArrayList<>();
+    for (int i = 0; i < eventList.size(); i++) {
+      Event curEvent = eventList.get(i);
+      int startOfCurEvent = curEvent.getWhen().start();
+
+      boolean endTimeIsInclusive = "EOD".equals(curEvent.getTitle());
+      TimeRange timeBetweenEvents =
+          TimeRange.fromStartEnd(endOfEarlierEvent, startOfCurEvent, endTimeIsInclusive);
+      if (timeBetweenEvents.duration() >= request.getDuration()) {
+        openMeetingTimes.add(timeBetweenEvents);
+      }
+
+      // Only move reference points of end of last event if the cur event end point is later
+      // than the end of the event with the latest end point so far.
+      int endOfCurEvent = curEvent.getWhen().end();
+      if (endOfCurEvent >= endOfEarlierEvent) {
+        endOfEarlierEvent = endOfCurEvent;
+      }
+    }
+
+    return openMeetingTimes;
+  }
+
+  /**
+   * Returns a collection of all meeting times that will work for all attendees' schedules.
+   *
+   * <p>To determine the meeting times ({@code TimeRange}s) that works for all attendees, the event
+   * list is filtered and sorted so that {@code getMeetingTimes()} can determine all feasible
+   * meeting times. More specifically, filtered means that an event is removed if it does not
+   * contain at least one participant that is included in the meeting request. Both filtering and
+   * then sorting are accomplished using a stream of Event objects.
+   *
+   * @param events The Collection of events that are used to determine what periods of time that the
+   *     meeting can take place.
+   * @param request The Meeting request that contains the requirements for potential meetings.
+   * @return A Collection of the feasible meeting {@code TimeRange}s.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList =
+        events.stream()
+            .filter(
+                event -> eventParticipantInMeeting(
+                    new HashSet<String>(event.getAttendees()), request.getAttendees()))
+            .sorted(
+                new Comparator<Event>() {
+                  @Override
+                  public int compare(Event a, Event b) {
+                    return Long.compare(a.getWhen().start(), b.getWhen().start());
+                  }
+                })
+            .collect(Collectors.toList());
```

Comment:
This method of solving the problem is very functional programming style, pretty cool!

Reply:
Thanks! I had never heard about stream() until Longjun told me about it but I agree that it is pretty sick.